### PR TITLE
ipa-cert-fix: Don't hardcode the NSS certificate nickname

### PR DESCRIPTION
**Pull request by rcritten**
ipa-cert-fix: Don't hardcode the NSS certificate nickname

The nickname of the 389-ds certificate was hardcoded as
Server-Cert which failed if the user had installed a
third-party certificate using ipa-server-certinstall.

Instead pull the nickname from the DS configuration and
retrieve it based on that.

https://pagure.io/freeipa/issue/8600

Signed-off-by: Rob Crittenden <rcritten@redhat.com>

**NOTE**: I'm not sure that the test is completely worth it as this is a bit of an edge case. It is not a fast test at all.

### CODE REVIEW

**Comment by rcritten** [comment]
test_ipa_cert_fix.py::TestIpaCertFix failed with:

[Errno 2] No such file or directory: '/etc/pki/pki-tomcat/certs/subsystem.crt'

This is unrelated to my change. IIRC @flo-renaud commented on the same error in another PR but I can't seem to find it. I just seem to recall it had to do with moving system time around.

**Comment by flo-renaud** [comment]
@rcritten the temp commit is running on fedora 32 with pki-server 10.10.0 which has a bug on pki-server cert-fix (https://bugzilla.redhat.com/show_bug.cgi?id=1897120). The bug was fixed in 10.10.2 only.

pki team released 10.10.3 2 weeks ago on fc32/fc33, this PR may be a good opportunity to bump the Requires version to at least 10.10.2.

File: ipaserver/install/ipa_cert_fix.py [hunk 1]
```diff
@@ -203,9 +203,12 @@ def expired_ipa_certs(now):
...
 
     # LDAPS
-    ds_dbdir = dsinstance.config_dirname(realm_to_serverid(api.env.realm))
+    serverid = realm_to_serverid(api.env.realm)
+    ds = dsinstance.DsInstance(realm_name=api.env.realm)
+    ds_dbdir = dsinstance.config_dirname(serverid)
+    ds_nickname = ds.get_server_cert_nickname(serverid)
     db = NSSDatabase(nssdir=ds_dbdir)
-    cert = db.get_cert('Server-Cert')
+    cert = db.get_cert(ds_nickname)
```
Comment (flo-renaud) [inline ipaserver/install/ipa_cert_fix.py:211]: There is another reference to 'Server-Cert' in the method `install_ipa_certs` that needs to be fixed.

**Review by flo-renaud — COMMENTED** [review]
Hi @rcritten
ipa-cert-fix doesn't seem to work properly with another NSS cert nickname. When I tried ipa-cert-fix post-expiration date, the new cert was put in the NSS db but appears with ',,' trust flags (as if the key was missing):
```
# certutil -L -d /etc/dirsrv/slapd-IPA-TEST/

Certificate Nickname                                         Trust Attributes
                                                             SSL,S/MIME,JAR/XPI

IPA.TEST IPA CA                                              CT,C,C
CN=CA,O=Example Organization                                 CT,C,C
CN=server.ipa.test,O=Example Organization                    ,,
```
and dirsrv fails to restart (`ERR - slapd_extract_key - Failed to extract keys for CN=server.ipa.test,O=Example Organization.`) although the key seems to be present in the DB:
```
# certutil -K -d /etc/dirsrv/slapd-IPA-TEST/ -f /etc/dirsrv/slapd-IPA-TEST/pwdfile.txt
certutil: Checking token "NSS Certificate DB" in slot "NSS User Private Key and Certificate Services"
< 0> rsa      1ba1fe7d0c8d8f5b2032c14e07a20a2b01f5cf41   (orphan)
< 1> rsa      fb44ec02cc53bdc17fb5585969c0575d3b4cd442   CN=server.ipa.test,O=Example Organization
```

File: freeipa.spec.in [hunk 2]
```diff
@@ -132,7 +132,7 @@
 %else
 # New KRA profile, ACME support
 # https://pagure.io/freeipa/issue/8545
-%global pki_version 10.10.0-2
+%global pki_version 10.10.3-1
```
[the remark this answers is not in the document]
Comment (rcritten) [reply freeipa.spec.in:135]: Ok will update.

**Comment by rcritten** [comment]
Very strange about the key. I suspect we need to ensure that any certs being fixed are issued by IPA first.

**Comment by rcritten** [comment]
Fixed the test failure. It was an issue on my end detecting whether a cert was IPA-issued.

**Comment by flo-renaud** [comment]
Hi @rcritten 
I tested with a LDAP cert issued by IPA CA but with a different nickname, the fix works well.
For certs issued by 3rd party CA, I still don't understand why the renewal fails but that's a fact, it doesn't work :(
I think the same check should also be added for the KDC cert if not issued by IPA.

**Comment by rcritten** [comment]
@flo-renaud my guess is it's failing because without the DS cert nothing can work. What I'd suggest is issuing 3rd party certs good for like 5 years so they are still valid when the other IPA certs need renewal. I'm going to test that locally.

**Comment by rcritten** [comment]
I've confirmed it. It works if I have 3rd party certs that expire after the IPA certs *and* dirsrv, httpd and krb5kdc are all running when I execute ipa-cert-fix (I'm not 100% sure that krb5kdc matters but I started it even though IPA issued its cert).

pki-tomcatd requires dirsrv at least in order to start properly. So ipa-cert-fix may renew the certs ok but will then fail trying to start the CA due to expired certs.

So in short: it's complicated.

I can think of three choices we can make:

1. Do nothing and let the end-user deal with it. It wasn't exactly obvious to me why tomcat didn't start and I only figured it out after poking at the certs to discover that my config tweak hadn't extended my 3rd party certs from 1 to 4 years.
2. While collecting the certs if its third party then also check validity and do...something?
3. Something else?

#2 seems doable but I don't know what we should do or recommend. Force them to renew their 3rd party certs first I guess?

**Comment by rcritten** [comment]
I chose option 2 and display the non-IPA renewed certificates so the user can manage it themselves. I also cleaned up the output a little and added a "Restarting IPA" message because it seemed like "Becoming renewal master" was hanging otherwise.

I tested with:

* All IPA-issued certificates
* Expired IPA + expired 3rd party HTTP and LDAP certificates
* Expired IPA + non-expired third party HTTP and LDAP certificates

**Comment by flo-renaud** [comment]
Hi @rcritten 
Thanks for the update, WFM.
- [x] all certs IPA-issued, already expired, and LDAP certificate with a different name: `ipa-cert-fix` succeeds
- [x] expired IPA certs, HTTP/LDAP/KDC 3rd part and still valid: `ipa-cert-fix` succeeds
- [x] valid IPA certs, HTTP/LDAP/KDC 3rd part and expired: `ipa-cert-fix` prints "Nothing to do" and exits with 0. Consistent with the man page that reads `This tool cannot renew certificates signed by external CAs`.
- [x] expired IPA certs, HTTP/LDAP/KDC 3rd part and expired: `ipa-cert-fix` prints the certs that will NOT be renewed, and renews the other ones.

You can drop the temp commit, and I will ACK when PRCI completes

**Comment by rcritten** [comment]
Dropped temp commit

**Comment by flo-renaud** [comment]
master:

* 7cfd44d395da0d499c0569b6db2cabd9fedb488d ipa-cert-fix: Don't hardcode the NSS certificate nickname
* be4195cfaf24341d0e9f172f3cd9b39ef14d5783 ipatests: test third-party 389-ds cert with ipa-cert-fix
* 146db28e50efa7a111468faf07e24ef3166c3117 Set pki-core dependency to 10.3.3 for pki-server cert-fix bug
* dea2b8af9c35708e98b8512e86fedfc95483b025 Don't renew non-IPA issued certs in ipa-cert-fix
